pypesto/objective/history.py

```diff
@@ -1371,148 +1412,179 @@ def finalize(self, message: str = None, exitflag: int = None):
+        ----------
+        x: A single parameter vector.
+
+        Returns
+        -------
+        admissible: Whether the point fulfills the problem requirements.
+        """
+        return np.all(x <= self.ub) and np.all(x >= self.lb)
+
+    def _get_optimal_point_from_history(self) -> ResultDict:
+        """Extract optimal point from `self.history`."""
+        result = {}
+
         # get indices of admissible trace entries
         # shape (n_sample, n_x)
         xs = np.asarray(self.history.get_x_trace())
         ixs_admit = [ix for ix, x in enumerate(xs) if self._admissible(x)]
 
-        # we prioritize fval over chi2 as fval is written whenever possible
+        if len(ixs_admit) == 0:
+            # no admittable indices
+            return {key: None for key in OptimizerHistory.MIN_KEYS}
+
+        # index of minimum of fval values
         ix_min = np.nanargmin(self.history.get_fval_trace(ixs_admit))
-        # np.argmin returns ndarray when multiple minimal values are found, we
-        # generally want the first occurrence
+        # np.argmin returns ndarray when multiple minimal values are found,
+        #  we want the first occurrence
         if isinstance(ix_min, np.ndarray):
             ix_min = ix_min[0]
         # select index in original array
         ix_min = ixs_admit[ix_min]
 
-        for var in ['fval', 'chi2', 'x']:
-            self.extract_from_history(var, ix_min)
-            if var == 'fval':
-                self.fval_min = float(self.fval_min)
-
-        for var in ['res', 'grad', 'sres', 'hess']:
-            if not getattr(self.history.options, f'trace_record_{var}'):
-                continue  # var not saved in history
-            # first try index of optimal function value
-            if self.extract_from_history(var, ix_min):
-                continue
-            # gradients may be evaluated at different indices, therefore
-            #  iterate over all and check whether any has the same parameter
-            #  and the desired field filled
-            # for res we do the same because otherwise randomly None
-            #  (TODO investigate why, but ok this way)
-            for ix in reversed(range(len(self.history))):
-                if not np.allclose(self.x_min, self.history.get_x_trace(ix)):
+        # fill in parameter and function value from that index
+        for var in (X, FVAL, RES):
+            val = getattr(self.history, f'get_{var}_trace')(ix_min)
+            if val is not None and not np.all(np.isnan(val)):
+                result[var] = val
+            # convert to float if var is FVAL to be sure
+            if var == FVAL:
+                result[var] = float(result[var])
+
+        # derivatives may be evaluated at different indices, therefore
+        #  iterate over all and check whether any has the same parameter
+        #  and the desired field filled
+        for var in (GRAD, HESS, SRES):
+            for ix in range(len(self.history)):
```

Comment:
this piece of code occurs multiple times, maybe add method that takes `var` as argument?